File: py/kson.py
#!/usr/bin/env python
# coding: utf-8
#
# Copyright (c) 2013, Manuel Barkhau <mbarkhau@gmail.com>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# CONSEQUENTIAL DAMAGES LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
import os
import re
import json
import string
import collections
from base64 import b64encode
from datetime import datetime
from calendar import timegm

default_json_dumps = json.dumps
default_json_loads = json.loads

try:
    import ujson
    json_loads = ujson.loads

    def json_dumps(*args, **kwargs):
        kwargs['encode_html_chars'] = False
        kwargs['ensure_ascii'] = False
        return ujson.dumps(*args, **kwargs)
except ImportError:
    json_dumps = default_json_dumps
    json_loads = default_json_loads


DECODERS = {}
ENCODERS = {}

CODEC_FACTORIES = {}
SCHEMAS = {}

_verbose = False


def _plain_id(id):
    return id[2:] if id and id[0] == u"[" else id


def baseN(num, b=36, digits=string.digits + string.ascii_letters):
    if num == 0:
        return digits[0]

    if num < 0:
        return "-" + baseN(-num, b, digits)

    d = digits[num % b]
    return baseN(num // b, b, digits).lstrip(digits[0]) + d


## codec management


def add_codec(codec_or_name):
    """decorator to register a kson codec factory function"""
    if not isinstance(codec_or_name, basestring):
        # assume the argument is the factory function
        codec = codec_or_name
        name = codec.__name__
        CODEC_FACTORIES[name] = codec
        return codec

    name = codec_or_name

    def dec(codec):
        CODEC_FACTORIES[name] = codec
        return codec

    return dec


def coder_fn(coders):
    def coder(val):
        for c in coders:
            val = c(val)
        return val
    return coder


CODEC_RE = re.compile(r"(?:\\.|[^\|])+")
ARGS_RE = re.compile(r"(?:\\.|[^:])+")


def init_codec(meta_id):
    if not meta_id or meta_id in ENCODERS:
        return

    metas = CODEC_RE.findall(meta_id)

    encoders = []
    decoders = []

    for meta in metas:
        args = ARGS_RE.findall(meta)
        codec_id = args[0]
        if codec_id not in ENCODERS:
            return

        args = args[1:]
        for j, arg in enumerate(args):
            args[j] = arg.replace(r"\\", "")

        encoders.append(ENCODERS[codec_id](args))
        decoders.append(DECODERS[codec_id](args))

    ENCODERS[meta_id] = coder_fn(encoders)
    DECODERS[meta_id] = coder_fn(reversed(decoders))


def init_codecs(schema):
    for meta_id in schema['meta']:
        init_codec(_plain_id(meta_id))


# default codecs


@add_codec('suffix')
def suffix_codec(args):
    suffix = args[0]

    def encoder(val):
        if not val.endswith(suffix):
            raise ValueError("Expected %s to have suffix %s" % (val, suffix))
        return val.rsplit(suffix, 1)[0]

    def decoder(raw):
        return raw + suffix

    return encoder, decoder


@add_codec('prefix')
def prefix_codec(prefix):
    def encoder(val):
        if not val.startswith(prefix):
            raise ValueError("Expected %s to have prefix %s" % (val, prefix))
        return val.replace(prefix, "", 1)

    def decoder(raw):
        return prefix + raw

    return encoder, decoder


@add_codec('enum')
def enum_codec(vals):
    def encoder(val):
        return vals.index(val)

    def decoder(raw):
        return vals[raw]

    return encoder, decoder


@add_codec('date')
def date_codec(args):
    def encoder(val):
        return baseN(timegm(val.utctimetuple()) * 1000, 36)

    def decoder(val):
        return datetime.fromtimestamp(int(val, 36) / 1000.0)

    return encoder, decoder


## schema


def add_schema(schema):
    if isinstance(schema, basestring):
        schema = loads(schema)

    if not schema:
        raise ValueError("Invalid Schema: " + str(schema))

    if 'id' not in schema:
        raise ValueError("Invalid Schema: Missing field 'id'")
    if 'fields' not in schema:
        raise ValueError("Invalid Schema: Missing field 'fields'")
    if 'meta' not in schema:
        raise ValueError("Invalid Schema: Missing field 'meta'")

    assert len(schema['fields']) == len(schema['meta'])

    if _verbose:
        print "Adding new schema: ", schema['id']
        print "\t", schema['fields']
        print "\t", schema['meta']

    init_codecs(schema)
    SCHEMAS[schema['id']] = schema


add_schema({
    'id': 'schema',
    'fields': ['id', 'fields', 'meta'],
    'meta': [0, "[]", "[]"]
})


def load_schemas(fp_or_filename):
    if isinstance(fp_or_filename, basestring):
        with open(fp_or_filename, 'r') as f:
            data = f.read()
    else:
        data = fp_or_filename.read()
    return loads_schemas(data)


def loads_schemas(schema_data):
    schemas = loads(schema_data)
    if isinstance(schemas, list):
        for s in schemas:
            add_schema(s)
    else:
        add_schema(schemas)


## (de)serialization


def _process_val(val, meta_id, recurse, coders):
    if val is not None:
        p_meta_id = _plain_id(meta_id)
        if p_meta_id in SCHEMAS:
            return recurse(val, meta_id, True)

        if p_meta_id in coders:
            coder = coders[p_meta_id]

            if p_meta_id == meta_id:
                return coder(val)

            val = val[:]  # copy so argument isn't modified
            for i, v in enumerate(val):
                val[i] = coder(v)
    return val


def dumps(data, schema_id, is_recurse=False, *args, **kwargs):
    is_array = schema_id[0] == u"["
    schema_id = schema_id[2:] if is_array else schema_id
    schema = SCHEMAS[schema_id]
    fields = schema['fields']
    meta = schema['meta']
    result = []

    if is_array and not isinstance(data, collections.Iterable):
        raise ValueError("Schema specifies array, got %s" % type(data))

    if not is_array and not isinstance(data, collections.Mapping):
        raise ValueError("Schema specifies object, got %s" % type(data))

    if not is_recurse:
        result.append("[]" + schema_id if is_array else schema_id)

    if not is_array:
        data = [data]

    for obj in data:
        for field, meta_id in zip(fields, meta):
            result.append(_process_val(
                obj.get(field, None), meta_id, dumps, ENCODERS
            ))

    if is_recurse:
        return result

    if args or kwargs:
        return default_json_dumps(result, *args, **kwargs)
    return json_dumps(result)


def loads(data, schema_id=None, is_recurse=False):
    data = json_loads(data) if isinstance(data, basestring) else data
    if not isinstance(data, list) or len(data) == 0:
        return data
    if not (schema_id or isinstance(data[0], basestring)):
        return data

    data_start = 0
    if not schema_id:
        schema_id = data[0]
        data_start = 1

    is_array = schema_id[0] == u"["
    if is_array:
        schema_id = schema_id[2:]
        result = []

    schema = SCHEMAS[schema_id]
    fields = schema['fields']
    len_fields = len(fields)
    meta = schema['meta']
    obj = None

    for i, val in enumerate(data[data_start:]):
        j = i % len_fields
        if j == 0:
            if obj and is_array:
                result.append(obj)
            obj = {}

        meta_id = meta[j]

        obj[fields[j]] = _process_val(val, meta_id, loads, DECODERS)

    if not is_array:
        return obj

    if obj:
        result.append(obj)
    return result


## helpers for dealing with files


def dump(data, fp_or_filename, *args, **kwargs):
    data = dumps(data, *args, **kwargs)
    if isinstance(fp_or_filename, basestring):
        with open(fp_or_filename, 'w') as fp:
            fp.write(data)
    else:
        fp_or_filename.write(data)


def load(fp_or_filename, *args, **kwargs):
    if isinstance(fp_or_filename, basestring):
        with open(fp_or_filename, 'r') as f:
            data = f.read()
    else:
        data = fp_or_filename.read()
    return loads(data, *args, **kwargs)


## schema detection


def detect_codecs(data, schema_id):
    return 0


def compact_schemas(base_schema_id):
    def replace_schema(old_sid, new_sid):
        for schema in SCHEMAS.values():
            meta = schema['meta']
            for i, meta_id in enumerate(meta):
                if meta_id == old_sid:
                    meta[i] = new_sid
        del SCHEMAS[old_sid]

    done = False
    while not done:
        done = True
        for a_sid, a_schema in SCHEMAS.items():
            if not a_sid.startswith(base_schema_id):
                continue

            a_fields = a_schema['fields']
            a_meta = a_schema['meta']
            for b_sid, b_schema in SCHEMAS.items():
                if not b_sid.startswith(base_schema_id):
                    continue
                if a_sid == b_sid:
                    continue

                b_fields = b_schema['fields']
                b_meta = b_schema['meta']

                if a_fields == b_fields and a_meta == b_meta:
                    replace_schema(b_sid, a_sid)
                    done = False
                    break
            if not done:
                break


def detect_meta(obj, id_prefix, lvl):
    idx = 0
    meta_map = {}
    for field, value in obj.items():
        if isinstance(value, dict) or isinstance(value, list):
            meta_map[field] = detect_schemas(value, id_prefix, lvl + 1, idx)
            idx += 1
    return meta_map


def detect_schemas(data, id_prefix=None, lvl=0, idx=0):
    if id_prefix is None:
        id_prefix = "auto-schema-" + b64encode(os.urandom(6))
    schema_id = id_prefix + "-" + str(lvl) + "-" + str(idx)
    fields = set()
    meta_map = {}
    if isinstance(data, dict):
        fields.update(data.keys())
        meta_map.update(detect_meta(data, id_prefix, lvl))
    elif isinstance(data, list):
        for obj in data:
            if isinstance(obj, dict):
                fields.update(obj.keys())
                meta_map.update(detect_meta(obj, id_prefix, lvl))
    else:
        raise ValueError("Top level must be a dict or list")

    if len(fields) == 0:
        if isinstance(data, list):
            return "[]"
        return 0

    fields = sorted(list(fields))

    add_schema({
        'id': schema_id,
        'fields': fields,
        'meta': [meta_map.get(f, 0) for f in fields]
    })

    if lvl == 0:
        compact_schemas(id_prefix)
        detect_codecs(data, schema_id)

    if isinstance(data, list):
        return "[]" + schema_id
    return schema_id

File: js/kson.js
/**
 * Copyright (c) 2013, Manuel Barkhau <mbarkhau@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * CONSEQUENTIAL DAMAGES LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *
 * KSON: Keyless Schemafied Object Notation
 *
 * A serialization format with two goals in mind:
 *	1. Easily parsable using only a small javascript library.
 *	2. Reduce serialized size compared to JSON.
 *
 * 1. is accomplished by using the (comparativly fast) JSON parse/stringify
 *	functions, thus reducing the task of KSON to packing/unpacking the values
 *	according to a predefined schema. This also greatly reduces the size of
 *	the library which is downloaded to the browser, compared with other
 *	serialization formats (gzipped KSON is < 1K).
 * 2. is accomplished by eliminating often redundant keys and by extending
 *	native JSON datatypes with an encoding/decoding mechanism. Codecs allow
 *	for a compact serialized data representation.
 */
(function (global, module) {

"use strict";

var SCHEMAS = {
		// default schema (meta) schema
		// needed to parse schemas in KSON format
		schema: {
			id: 'schema',
			fields: ['id', 'fields', 'meta'],
			meta: [0, "[]", "[]"]
		}
	},
	// Encoder/decoder functions are Factory functions that are
	// initialized when a schema is loaded. They return function
	// closures that are used for the actual encoding/decoding.
	CODERS = {
		'enum': function(args) {
			var i = args.unshift(0),
				arg_indexes = {};
			for (;i--;) {
				arg_indexes[args[i]] = i;
			}
			return function (val, enc) {
				// If the value is not enumerated in the definition
				// the value is used as is. This is an issue if the
				// field value is an int within the range of the
				// enumerated values as it will be decoded incorrectly.
				return (enc ?
					arg_indexes[val] || val:
					args[val] || val
				);
			};
		},
		'prefix': function (args) {
			var prefix = new RegExp("^" + args[0]);
			return function (val, enc) {
				return enc ? val.replace(prefix, "") : args[0] + val;
			};
		},
		'suffix': function (args) {
			var suffix = new RegExp(args[0] + "$");
			return function (val, enc) {
				return enc ? val.replace(suffix, "") : val + args[0];
			};
		}
		// TODO: add build options to include various coders
		// I don't think this one will be used often enough to warrant
		// inclusion in the default build.

		// 'offsetint': function (args) {
		// 	var offset = parseInt(args[0], 10);
		// 	return function (val, enc) {
		// 		return enc ? val - offset : val + offset;
		// 	};
		// },
	},
	// This holds the results of calls to 'coder_fn' based
	// on configurations used by schemas.
	meta_coders = {};


function plain_id(id) {
	return id[0] == "[" ? id.slice(2) : id;
}

function base_coder(args, codec_id) {
	return function (val, enc) {
		return (
			val == null ?
				val
			: codec_id == 'bool' ?
				(enc ? val && 1 || 0			: val && true || false)
			// Numbers less than 100 000 000 are shorter or equivalent
			// in length to a base 36 encoded number in base 10, because
			// base 36 requires two extra bytes for the quotes. One could
			// do a decoder which will parse strings as base36 and numbers
			// are just returned as is, and an encoder which will only
			// .toString(36) for values > 10^8
			//: codec_id == 'int36' ?
			//	(enc ? val.toString(36) 		: parseInt(val, 36))
			: codec_id == 'date' ?
				(enc ? Math.round(val / 1000) 	: new Date(val * 1000))
			: codec_id == 'isodate' ?
				(enc ? val.toISOString() 		: new Date(Date.parse(val)))
			: val
		);
	};
}


// Parses the coder definitions
function init_meta_coders(schema) {
	function init_meta_coder(meta_id) {
		function coder_fn(coders) {
			return function (val, dir) {
				// dir < 0 : encoed
				// dir > 0 : decode
				var len = coders.length,
					cur = dir > 0 ? 0 : len - 1,
					end = dir > 0 ? len : -1

				for (; cur != end; cur += dir) {
					val = coders[cur](val, dir != 1);
				}
				return val;
			};
		}

		function parse_coders() {
			// multiple coders for same field are separated by "|" (pipe characters)
			var metas = meta_id.match(/(\\.|[^\|])+/g),
				coder_id, args, coders = [],
				i = metas.length, j;

			for (; i--;) {
				coder_id = metas[i];
				args = coder_id.match(/(\\.|[^:])+/g);
				coder_id = args[0];

				// strip off coder id leaving only coder arguments
				args = args.slice(1);
				// remove escape chars
				for (j = args.length; j--;) {
					args[j] = args[j].replace("\\", "");
				}

				coders.push((CODERS[coder_id] || base_coder)(args, coder_id));
			}
			return coders;
		}
		if (
			meta_id 				// 0 placeholder (no coder)
			&& !SCHEMAS[meta_id]	// schema/not a coder
		) {
			meta_coders[meta_id] = coder_fn(parse_coders(meta_id));
		}
	}

	var meta = schema['meta'] = schema['meta'] || [],
		i = schema['fields'].length;
	for (;i--;) {
		(meta[i] = meta[i] || 0) && init_meta_coder(plain_id(meta[i]));
	}
}

function process_val(val, meta_id, recurse, dir) {
	// dir < 0 : encode, dir > 0 : decode
	var i, p_meta_id = plain_id(meta_id),
		coder = meta_coders[p_meta_id];

	if (coder && p_meta_id != meta_id) {
		val = val.slice(0); // copy so original reference isn't modified
		for (i = val.length; i--;) {
			val[i] = coder(val[i], dir);
		}
		return val;
	}
	return (
		val == null			? val :
		SCHEMAS[p_meta_id]	? recurse(val, meta_id, 1) :
		coder				? coder(val, dir) : val
	);
}

function stringify(data, schema_id, is_subarray) {
	var p_schema_id = plain_id(schema_id),
		i = 0, j, obj,
		result = [];

	if ((data instanceof Array) && schema_id[0] != "[") {
		schema_id = "[]" + schema_id;
	}
	if (!is_subarray) {
		result[0] = schema_id;
	}
	if (p_schema_id == schema_id) {
		data = [data];
	}
	var schema = SCHEMAS[p_schema_id],
		fields = schema['fields'],
		data_length = data.length,
		fields_length = fields.length;

	for (; i < data_length; i++) {
		obj = data[i];
		for (j = 0; j < fields_length; j++) {
			result.push(process_val(
				obj[fields[j]], schema['meta'][j], stringify, -1
			));
		}
	}
	return (is_subarray) ? result : JSON.stringify(result);
}

function parse(raw, schema_id, recurse) {
	var data = (typeof raw == 'string') ? JSON.parse(raw) : raw,
		result = [], tmp_obj,
		i = recurse ? 0 : 1, j;

	if (!schema_id) {
		schema_id = data[0];
	}

	var p_schema_id = plain_id(schema_id),
		schema = SCHEMAS[p_schema_id],
		fields = schema['fields'],
		data_length = data.length,
		fields_length = fields.length;

	for (; i < data_length; i += fields_length) {
		tmp_obj = {};
		for (j = 0; j < fields_length; j++) {
			tmp_obj[fields[j]] = process_val(
				data[i + j], schema['meta'][j], parse, 1
			);
		}

		if (schema_id != p_schema_id) {
			result.push(tmp_obj);
		} else {
			return tmp_obj;
		}
	}
	return result;
}

/**
 * Initialize a schema or an array of schemas from
 *  - a javascript object
 *  - a JSON encoded schema string
 *  - a KSON encoded schema string.
 */
function addSchema(schema) {
	// TODO: with circular schema definitions it may be
	//		better to insert the prototype automatically
	//		or at least raise an explicit error.
	if (typeof schema === 'string') {
		schema = parse(schema);
	}
	if (schema instanceof Array) {
		// Order is important here
		// (independent schemas must be loaded first)
		for (var i = schema.length; i--;) {
			addSchema(schema[i]);
		}
	} else {
		SCHEMAS[schema['id']] = schema;
		init_meta_coders(schema);
	}
}

// exports
global.KSON = module.exports = {
	coders: CODERS,
	addSchema: addSchema,

	parse: parse,
	stringify: stringify
};

})(this, typeof module == "undefined" ? {} : module);
